src/mpid/ch4/include/mpidpre.h

```diff
@@ -227,45 +227,57 @@ struct MPIDI_CH4U_win_lock {
     uint16_t type;
 };
 
-struct MPIDI_CH4U_win_queue {
+typedef struct MPIDI_CH4U_win_lock_recvd {
     struct MPIDI_CH4U_win_lock *head;
     struct MPIDI_CH4U_win_lock *tail;
-};
+    int type;                   /* current lock's type */
+    unsigned count;             /* count of granted locks (not received) */
+} MPIDI_CH4U_win_lock_recvd_t;
 
-typedef struct MPIDI_CH4U_win_lock_info {
-    unsigned peer;
-    int lock_type;
-    struct MPIR_Win *win;
-    volatile unsigned done;
-} MPIDI_CH4U_win_lock_info;
+typedef struct MPIDI_CH4U_win_target_sync_lock {
+    volatile unsigned locked;   /* locked == 0 or 1 */
```

Comment:
If multiple threads are sharing this `locked` variable without taking a lock (say, pthread mutex), then volatile alone isn't enough, because hardware might also optimize the instructions, so you'd need a memory barrier in such a case. If a proper lock primitive (e.g. pthread mutex) is used, then neither volatile or memory barrier is needed in your code because lock primitive will do it on behalf of you (cf. http://stackoverflow.com/a/3208140)

I think the current MPICH code falls into the latter case with a lock, at least in the global lock mode.